### PR TITLE
CLI remote connections

### PR DESCRIPTION
**Pull request by macneale4**
Enable the connection of SQL enabled CLI commands to remote servers with addition of three flags:
* --host - The server to connect to. Using this flag forces the use of a remote connection no matter what.
* --port - The port to use. Defaults to 3306. No effect when --host is not used.
* --no-tls - Disable the use of SSL. This is required to connect to your local host.

Also special cased the `sql` command to allow starting without a DB. Enables connecting to an empty repo so you can `CREATE DATABASE FOO`.

### CODE REVIEW

**Comment by nicktobey** [comment]
Initial thought before actually reading any of the code: Can using --port without --host be an error instead of the --port flag having no effect?

File: go/cmd/dolt/dolt.go [hunk 1]
```diff
@@ -558,11 +558,29 @@ func runMain() int {
 	return res
 }
 
+// buildLateBinder builds a LateBindQueryist for which is used to obtain the Querist used for the length of the
```
Comment (nicktobey) [inline go/cmd/dolt/dolt.go:561]: s/Querist/Queryist

File: go/cmd/dolt/commands/show.go [hunk 2]
```diff
@@ -112,6 +112,10 @@ func (cmd ShowCmd) ArgParser() *argparser.ArgParser {
 	return ap
 }
 
+func (cmd ShowCmd) RequiresRepo() bool {
+	return false
```
Comment (nicktobey) [inline go/cmd/dolt/commands/show.go:116]: This (and the change to diff) feel unrelated to enabling remote connections and probably belong in separate PR.

File: integration-tests/bats/cli-hosted.bats [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+# Honestly, this test was created in haste, and because multiple parts of the default setup infrastructure don't work
+# for this use case.
+
+setup() {
+  setup_no_dolt_init
+
+  mkdir emptyDB
+  cd emptyDB
+
+  # Unusually situation - this is the only test that I know of where we want to start a server completely
```
Comment (nicktobey) [inline integration-tests/bats/cli-hosted.bats:19]: s/Unusually/Unusual

File: integration-tests/bats/cli-hosted.bats [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  run dolt $TLS --host $HST --port $PRT --user bogus sql -q "create database bats_test_cli_hosted"
+  [ "$status" -eq 1 ]
+  [[ $output =~ "User not found 'bogus'" ]] || false
+}
+
+# This test will not work if you change DOLT_CLI_USER above. Be aware if you run against a hosted instance.
+@test "cli-hosted: bogus password rejected" {
+  run dolt $TLS --host $HST --port $PRT --user dolt --password bogus sql -q "create database bats_test_cli_hosted"
+
+  echo "-------------------"
```
Comment (nicktobey) [inline integration-tests/bats/cli-hosted.bats:99]: Is this leftover from debugging?

File: go/cmd/dolt/commands/show.go (hunk 2 above)
Comment (nicktobey) [reply go/cmd/dolt/commands/show.go:116]: I see now that the bats test uses them. These should _probably_ be flipped on in a separate RP and then merged into this one... but I don't think it's a big deal either way.

**Review by nicktobey — APPROVED** [review]
Very cool. I only have a couple of nits.

File: go/cmd/dolt/dolt.go (hunk 1 above)
Comment (stephkyou) [inline go/cmd/dolt/dolt.go:561]: spelling "Queryist"

**Review by stephkyou — APPROVED** [review]
LGTM!

File: go/cmd/dolt/commands/show.go (hunk 2 above)
Comment (macneale4) [reply go/cmd/dolt/commands/show.go:116]: They should have been enabled in the original migration. Making another PR seems like overkill.

File: integration-tests/bats/cli-hosted.bats (hunk 4 above)
Comment (macneale4) [reply integration-tests/bats/cli-hosted.bats:99]: doh.

**Comment by macneale4** [comment]
> Initial thought before actually reading any of the code: Can using --port without --host be an error instead of the --port flag having no effect?

I was going to punt on that, but I went ahead and did it

**Comment by nicktobey** [comment]
🚢

**Comment by macneale4** [comment]
Enable the connection of SQL enabled CLI commands to remote servers with addition of three flags:

--host - The server to connect to. Using this flag forces the use of a remote connection no matter what.
--port - The port to use. Defaults to 3306. No effect when --host is not used.
--no-tls - Disable the use of SSL. This is required to connect to your local host.
Also special cased the sql command to allow starting without a DB. Enables connecting to an empty repo so you can CREATE DATABASE FOO.

Related: https://github.com/dolthub/dolt/issues/3922
